tower-http/src/classify_response.rs

```diff
@@ -0,0 +1,82 @@
+use http::{Response, StatusCode};
+use std::fmt;
+
+/// Classify requests results as either success or failure.
+///
+/// Not all services can rely entirely on the status code to determine if its a success or failure.
+/// gRPC for example has its own way of communicating that and always uses `200 OK`.
+pub trait ClassifyResponse<Body, Err> {
+    /// The type used to classify successful responses.
+    type OkClass;
+
+    /// The type used to classify failed responses.
+    type ErrClass;
```

Comment:
nit: I might change this naming to `SuccessClass` and `FailureClass`, or even just `Success` and `Failure` — using `Ok` and `Err` feels like it kind of conflates these with whether the `Result` passed into the classifier is `Ok` or `Err`, but an `Ok` result might still be a _failure_ at the HTTP level, if it contains a 5xx status code.

Reply:
also, do we really need a class type for successful responses? if a response succeeded, do we expect users to want to classify that success into different kinds of success, or is it just "not an error"? IMO, it would _probably_ be fine for the success side of this to be `()`...but maybe leaving it as a type param is more future-proof? i could be convinced either way...

Reply:
`SuccessClass` and `FailureClass` sound better 👍

Regarding yielding a value from the classification that was a way to get a status code to the thing that emits the events.